### PR TITLE
nixStatic: fix darwin build

### PR DESCRIPTION
**Pull request by Mic92**
## Things done

~~Depends on https://github.com/NixOS/nixpkgs/pull/355350~~

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/tools/package-management/nix/common.nix [hunk 1]
```diff
@@ -174,6 +189,8 @@ self = stdenv.mkDerivation {
     (darwinMinVersionHook "10.13")
   ];
 
+  NIX_LDFLAGS = lib.optionalString (stdenv.hostPlatform.isDarwin && stdenv.hostPlatform.isStatic)
+      "-framework CoreFoundation -framework SystemConfiguration";
```
Comment (roberth) [inline pkgs/tools/package-management/nix/common.nix:193]: Other packages use `buildInputs` for this, and actual `CoreFoundation` (etc.) packages.

File: pkgs/tools/package-management/nix/common.nix [hunk 2]
```diff
@@ -163,6 +164,20 @@ self = stdenv.mkDerivation {
...
+      mkdir -p $out/lib/pkgconfig
+      cat > $out/lib/pkgconfig/iconv.pc <<EOF
+      prefix=${iconv}
+
+      Name: iconv
+      Description: GNU iconv
+      Version: ${lib.getVersion iconv}
+      Libs: -liconv
+      EOF
+    '')
```
Comment (roberth) [inline pkgs/tools/package-management/nix/common.nix:179]: Surely there's a better way to fix this, but if this gets the job done, I'm not opposed.

I guess you could use `writeTextFile` here, but that's obviously not the problem.

File: pkgs/tools/package-management/nix/common.nix (hunk 1 above)
Comment (Mic92) [reply pkgs/tools/package-management/nix/common.nix:193]: I don't think adding the package will actually cause the build to link against the framework. The issue I am adding these is that I see missing symbols when linking libcurl.

**Comment by Ericson2314** [comment]
Yeah if we can just make sure `iconv` has a pkg-config file on `staging` after this, to replace this, that would be much better.

File: pkgs/tools/package-management/nix/common.nix (hunk 1 above)
Comment (emilazy) [reply pkgs/tools/package-management/nix/common.nix:193]: Generally these forced linkages mean something is broken with the build upstream, I think.

(Also the separate framework packages are deprecated compatibility stubs now and do nothing at all.)

**Review by emilazy — COMMENTED** [review]
Apple doesn’t ship a `.pc` file for `libiconv` so it seems like an issue of the Nix build system if it’s requiring it on macOS.

File: pkgs/tools/package-management/nix/common.nix [hunk 3]
```diff
@@ -163,6 +164,20 @@ self = stdenv.mkDerivation {
...
+    iconv
```
Comment (emilazy) [inline pkgs/tools/package-management/nix/common.nix:180]: Are you sure you want this? This is the `iconv(1)` binary. `libiconv` is already provided by the base SDK.

**Comment by emilazy** [comment]
This is an upstream non‐portable brokenness that causes problems with other platforms like musl too: https://github.com/libarchive/libarchive/issues/1819 https://github.com/libarchive/libarchive/issues/1766. AFAICT it’s wrong on glibc too, it just happens to work by accident somehow(?). It seems it was a regression introduced for MSYS2 support or something. I don’t think any major `libiconv` (Apple’s, glibc’s, …) ships a `.pc` file, so we probably shouldn’t ship one ourselves.

We should fix it by fixing the `.pc` file in `libarchive` as Homebrew does, invariant of platform: https://github.com/Homebrew/homebrew-core/blob/f8e9e8d4f30979dc99146b5877fce76be6d35124/Formula/lib/libarchive.rb.

The `-framework` stuff I’m not sure about, but it’s almost certainly something broken in Nix’s build system or in the `.pc` files of libraries it’s linking to.

**Comment by emilazy** [comment]
BTW this package has been broken since 2023 so I don’t think it’s any kind of blocker. Static builds are pretty niche on macOS. If we fix it, we should fix it the right way, not by hacking around bugs in other packages and Nix’s build system like this.

**Comment by Mic92** [comment]
> BTW this package has been broken since 2023 so I don’t think it’s any kind of blocker. Static builds are pretty niche on macOS. If we fix it, we should fix it the right way, not by hacking around bugs in other packages and Nix’s build system like this.

Well. Should we just disable it than? It would make my life as a nix maintainer easier.

**Comment by Mic92** [comment]
In the near time, I don't think I have the capacity to do the suggested refactorings. My goal was just to get a clean nixpkgs review for nix upgrade. In any case as a proper fix would take a staging cycle, I would suggest to disable static nix darwin builds for now so it doesn't waste build cycles on peoples machines / hydra.

**Pull request closed without merging**